frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -946,8 +948,22 @@ export default ({
           const newContractState = await sbp('chelonia/in/processMessage', serializedMessage, this.messageState.contract)
 
           if (!this.checkEventSourceConsistency(contractID)) return
-          Vue.set(this.messageState, 'contract', newContractState)
 
+          if (this.isInCypress) {
+            // NOTE: When the user's actions are very quick, that can logout before to save `readUntilMessageHash`,
+            //       we should save `readUntilMessageHash` before to update contract state.
+            //       This normally happens in Cypress, when user logs out just after sending a message.
+            const curMessages = newContractState.messages || []
+            if (addedOrDeleted === 'ADDED' && curMessages.length) {
+              const lastMessage = curMessages[curMessages.length - 1]
+              this.updateReadUntilMessageHash({
+                messageHash: lastMessage.hash,
+                createdHeight: lastMessage.height
+              })
+            }
+          }
```

Comment:
> What issue is this fixing, and which Cypress test would be failing without it?

This was a very picky situation which kept making Cypress to be failed. Let me explain.
Once we receive the message event(e.g., `'gi.contracts/chatroom/addMessage'`), we process the event in this function and the `this.messageState` is updated. So the computed value `this.messages` would be updated.
After that, we save the `readUntil` message to the KV store by calling `this.updateReadUntilMessageHash` function.

Now, let's think about Cypress test scenario to logout as soon as to send a message. So the workflow is like the following.

- User sends a message
- Cypress waits for the message to be added (`pending` now)
- Cypress waits for the message status to be `sent` (This is when the event is received, processed and `this.messageState` is updated)
- (Cypress decides the `Sending a message` is finished, actually is.) Call cy.giLogout()

So the Cypress didn't wait for the `this.updateReadUntilMessageHash` to be called, and also for the invocations from the queue `UNREAD_MESSAGES_QUEUE`. The invocations from `UNREAD_MESSAGES_QUEUE` calls the following sbp function, so if the invocations were called after logout, it would break Cypress.
```
return sbp('chelonia/kv/set', ourIdentityContractId, KV_KEYS.UNREAD_MESSAGES, data, {
  encryptionKeyId: sbp('chelonia/contract/currentKeyIdByName', ourIdentityContractId, 'cek'),
  signingKeyId: sbp('chelonia/contract/currentKeyIdByName', ourIdentityContractId, 'csk'),
})
```
Adding `await sbp('okTurtles.eventQueue/queueEvent', UNREAD_MESSAGES_QUEUE, () => {})` in the `gi.actions/identity/logout` function, couldn't solve the problem because a new invocation could be added after we call the `gi.actions/identity/logout` function. When the user's actions are very quick, only in Cypress. Look at the following orders of function calls

- Update this.messageState in ChatMain.vue
- Call `gi.actions/identity/logout` function
- Call `await sbp('okTurtles.eventQueue/queueEvent', UNREAD_MESSAGES_QUEUE, () => {})`
- Call `this.updateReadUntilMessageHash` and add new invocation to `UNREAD_MESSAGES_QUEUE`

> Is it possible to fix this in some other way, perhaps by modifying the Cypress tests themselves and having them wait on something?

I did also tried to find another way to wait for this in Cypress, but couldn't find alternative solution which doesn't use `cy.wait()`.

Reply:
I have just had an inspiration that add exception handler in `fetchChatRoomUnreadMessages` and `saveChatRoomUnreadMessages` function. I will update the PR soon.

Reply:
@taoeffect, I will re-write the test scenario and remove the picky cases.